src/nupic/algorithms/Cells4.cpp

```diff
@@ -1880,206 +1892,373 @@ void Cells4::reset() {
   //}
 }
 
+struct self_t {
+  UInt nColumns;
+  UInt nCellsPerCol;
+  UInt nCells;
+  UInt activationThreshold;
+  UInt minThreshold;
+  UInt newSynapseCount;
+  UInt nIterations;
+  UInt nLrnIterations;
+  UInt segUpdateValidDuration;
+  Real initSegFreq;
+  Real permInitial;
+  Real permConnected;
+  Real permMax;
+  Real permDec;
+  Real permInc;
+  Real globalDecay;
+  bool doPooling;
+  bool checkSynapseConsistency;
+  bool ownsMemory;
+  bool resetCalled;
+  UInt pamLength;
+  UInt maxInfBacktrack;
+  UInt maxLrnBacktrack;
+  UInt maxSeqLength;
+  UInt learnedSeqLength;
+  Real avgLearnedSeqLength;
+  UInt maxAge;
+  UInt verbosity;
+  Int maxSegmentsPerCell;
+  Int maxSynapsesPerSegment;
+  Real avgInputDensity;
+  UInt pamCounter;
+};
+// NOTE:  ownsMemory cannot be saved in serialization.
+//        It indicates who currently allocated the buffers.
```

Comment:
Well...I need this for CSharp too because CSharp has a different heap (managed) than C++.  It cannot directly access C++'s heap space.  So to avoid a copy we need to do this.  I will try to make this a little cleaner however.